TypeCobol.LanguageServer/WorkspaceProjectStore.cs

```diff
@@ -0,0 +1,347 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using TypeCobol.Analysis;
+using TypeCobol.Compiler;
+using TypeCobol.Compiler.Directives;
+using TypeCobol.Compiler.File;
+using TypeCobol.Compiler.Text;
+using TypeCobol.LanguageServer.Context;
+using TypeCobol.Tools.APIHelpers;
+using TypeCobol.Tools.Options_Config;
+
+namespace TypeCobol.LanguageServer
+{
+    /// <summary>
+    /// A class representing a concept to store all WorkspaceProject instances known by the Workspace.
+    /// Internally it is a concurrent dictionary from WorkspaceProject Keys to WorkspaceProject instances.
+    /// </summary>
+    class WorkspaceProjectStore
+    {
+        /// <summary>
+        /// Exception thrown when project keys are duplicated.
+        /// </summary>
+        internal class DuplicatedProjectException : Exception
+        {
+            internal DuplicatedProjectException(string key) : base(key)
+            {
+            }
+        }
+
+        /// <summary>
+        /// Exception thrown when project key does not correspond to an existing WorkspaceProject.
+        /// </summary>
+        internal class UnknownProjectException : Exception
+        {
+            internal UnknownProjectException(string key) : base(key)
+            {
+            }
+        }
+
+        /// <summary>
+        /// The dictionary of WorkspaceProject Keys to WorkspaceProject instance.
+        /// </summary>
+        private readonly ConcurrentDictionary<string, WorkspaceProject> _workspaceProjects;
+        private readonly string _rootDirectoryFullName;
+        /// <summary>
+        /// A Dictionary to associate an Uri of a Document to its Context.
+        /// it represents all documents belonging to all workspace projects.
+        /// </summary>
+        private readonly ConcurrentDictionary<Uri, DocumentContext> _allOpenedDocuments;
```

Comment:
I think we should also have a dictionary giving for a document its project so it will avoid all iterations over the projects to find the one that the document is attached to.

Reply:
Also the WorkspaceProjectStore should keep a reference  to its parent Workspace so as to avoid passing workspace instances in some methods. I t simplifies the code / API and the association between WorkspaceProjectStore and Workspace is really easy to manage as it is created at construction time and never changes.

Reply:
Forget the remark about having another dictionary, the document-to-project association should be added directly into `DocumentContext` (as proposed in the new model).